api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java

```diff
@@ -646,6 +648,78 @@ public DbUser syncComplianceTrainingStatus(DbUser dbUser)
     }
   }
 
+  /** Syncs the current user's training status from Moodle. */
+  public DbUser syncComplianceTrainingStatus() throws org.pmiops.workbench.moodle.ApiException, NotFoundException {
+    return syncComplianceTrainingStatus(userProvider.get());
+  }
+
+  /**
+   * Updates the given user's training status from Moodle.
+   *
+   * <p>We can fetch Moodle data for arbitrary users since we use an API key to access Moodle,
+   * rather than user-specific OAuth tokens.
+   *
+   * <p>Using the user's email, we can get their badges from Moodle's APIs. If the badges are marked
+   * valid, we store their completion/expiration dates in the database. If they are marked invalid,
+   * we clear the completion/expiration dates from the database as the user will need to complete a
+   * new training.
+   */
+  public DbUser syncComplianceTrainingStatus(DbUser dbUser) throws org.pmiops.workbench.moodle.ApiException, NotFoundException {
+    // Skip sync for service account user rows.
+    if (isServiceAccount(dbUser)) {
+      return dbUser;
+    }
+
+    try {
+      Timestamp now = new Timestamp(clock.instant().toEpochMilli());
+      Map<String, BadgeDetails> userBadgesByName = complianceService.getUserBadgesByName(dbUser.getUsername());
+      if (userBadgesByName.containsKey(complianceService.getResearchEthicsTrainingField())) {
+        BadgeDetails complianceBadge = userBadgesByName.get(complianceService.getResearchEthicsTrainingField());
+        if(complianceBadge.getValid()) {
+          if (dbUser.getComplianceTrainingCompletionTime() == null) {
+            // The badge was previously invalid and is now valid.
+            dbUser.setComplianceTrainingCompletionTime(now);
+          } else if (!dbUser.getComplianceTrainingExpirationTime().equals(new Timestamp(complianceBadge.getDateexpire()))) {
+            // The badge was previously valid, but has a new expiration date (and so is a new training)
+            dbUser.setComplianceTrainingCompletionTime(now);
+          }
+          // Always update the expiration time.
+          dbUser.setComplianceTrainingExpirationTime(new Timestamp(complianceBadge.getDateexpire()));
+        }
+        else {
+          // The current badge is invalid or expired, the training must be completed or retaken.
+          dbUser.clearComplianceTrainingCompletionTime();
+          dbUser.clearComplianceTrainingExpirationTime();
+        }
+      }
+      else {
+        // There is no record of this person having taken the training.
+        dbUser.clearComplianceTrainingCompletionTime();
+        dbUser.clearComplianceTrainingExpirationTime();
+      }
+
+      return updateUserWithRetries(
+          u -> {
+            u.setComplianceTrainingExpirationTime(u.getComplianceTrainingExpirationTime());
```

Comment:
`dbUser` is `u` here. I just kind of cargo culted this block over from the last function.

Looking at `updateUserWithRetries` now, it applies the function passed to it to the user passed to it and also sets `lastModifiedTime` and enforces data access level changes as a side effect of aforementioned function.

These Moodle functions are the only place we do something like this - everywhere else we pass in a value unattached to the `dbUser` we're changing.  I think that, as originally intended, this function was mean to apply changes calculated elsewhere and we're just mutating `dbUser` so as to not have to think about nullity semantics

What do you think about something like:
```
Timestamp newComplianceTrainingCompletionTime = null;
Timestamp newComplianceTrainingExpirationTime = null;
{...}
if (badge valid) {
    if (badge was previously invalid and is now valid) {
        newComplianceTrainingCompletionTime = now
    } else if (old expiry time is not new expiry time) {
        newComplianceTrainingCompletionTime = now
    }
    newComplianceTrainingExpirationTime = badgeExpiryDate
}
// get rid of all the lines nulling out timestamps and get rid of those 'clear' functions
{...}
return updateUserWithRetries(u -> {
    u.setComplianceTrainingCompletionTime(newComplianceTrainingCompletionTime);
    u.setComplianceTrainingExpirationTime(newComplianceTrainingExpirationTime);
}, dbUser);

Reply:
👍 That approach seems much better, especially if it's more consistent with how we do it elsewhere.

Reply:
nit: This still feels more badgey than usery, and like the kind of thing that could go in its own little service. While we have a very wide data model for the user table, we don't have to make its service know everything about how to calculate those columns.

FWIW, there are a handful of things whose recalculation is triggered on every user "change", and that's one of the main motivations of the `withRetries()` stuff.

Reply:
I'd be ok with moving it, but I'd rather do so after deprecating the old call so that I don't have to either have the v1 and v2 in separate places or move the deprecated function.